crates/core/tedge_api/src/address.rs

```diff
@@ -0,0 +1,47 @@
+/// An address which could be either a target or source of messages
+///
+/// Nesting addresses allows to disambiguated between different kind of
+/// sources and the way they have arrived here.
+#[derive(Debug, Clone)]
+pub struct Address {
+    endpoint: EndpointKind,
+    source: Option<Box<Address>>,
+}
+
+impl Address {
+    pub fn new(endpoint: EndpointKind) -> Address {
+        Self {
+            endpoint,
+            source: None,
+        }
+    }
+
+    pub fn endpoint(&self) -> &EndpointKind {
+        &self.endpoint
+    }
+
+    /// Get the original source of an `Address`
+    pub fn origin(&self) -> &Address {
+        if let Some(source) = self.source.as_ref() {
+            source.origin()
+        } else {
+            self
+        }
+    }
+
+    pub fn add_new_step(&self, endpoint: EndpointKind) -> Self {
```

Comment:
I don't like the name of this function, as it does not describe its purpose particularly well.

Maybe something that describes the idea of "use this as a source for a newly created Address object". Maybe `derive_new`, but I don't like this particularly well either.